File: frontend/src/types/order.ts
/**
 * Order types - Story 3.3: Checkout Information & Payment Gateway
 * Extended for Story 4.2: Owner Order Board
 */

export type PaymentMethod = "cod" | "vnpay" | "momo" | "internal";
export type OrderStatus =
  | "pending"
  | "confirmed"
  | "in_progress"
  | "checked"
  | "shipped"
  | "delivered"
  | "cancelled"
  // Epic 10: new statuses
  | "pending_measurement"
  | "preparing"
  | "ready_to_ship"
  | "ready_for_pickup"
  | "in_production"
  | "renting"
  | "returned"
  | "completed";

export type PaymentStatus = "pending" | "paid" | "failed" | "refunded";

export interface ShippingAddress {
  province: string;
  district: string;
  ward: string;
  address_detail: string;
}

export type ServiceType = "buy" | "rent" | "bespoke";
export type SecurityType = "cccd" | "cash_deposit";

export interface RentalCheckoutFields {
  pickup_date: string;
  return_date: string;
  security_type: SecurityType;
  security_value: string;
}

export interface CreateOrderInput {
  customer_name: string;
  customer_phone: string;
  shipping_address: ShippingAddress;
  shipping_note?: string;
  payment_method: PaymentMethod;
  items: {
    garment_id: string;
    transaction_type: "buy" | "rent" | "bespoke";
    size?: string;
    start_date?: string;
    end_date?: string;
    rental_days?: number;
  }[];
  voucher_codes?: string[];
  rental_fields?: RentalCheckoutFields;
  measurement_confirmed?: boolean;
}

export interface OrderItemResponse {
  garment_id: string;
  garment_name: string;
  image_url: string | null;
  transaction_type: "buy" | "rent";
  size?: string | null;
  rental_days?: number | null;
  unit_price: number;
  total_price: number;
}

export interface OrderResponse {
  id: string;
  status: OrderStatus;
  payment_status: PaymentStatus;
  subtotal_amount: number;
  discount_amount: number;
  total_amount: number;
  applied_voucher_ids: string[];
  payment_method: PaymentMethod;
  payment_url?: string | null;
  customer_name: string;
  customer_phone: string;
  shipping_address: ShippingAddress | null;
  shipping_note?: string | null;
  is_internal?: boolean;
  items: OrderItemResponse[];
  created_at: string;
  tailor_info?: TailorInfoForCustomer[] | null;
  service_type?: ServiceType;
  deposit_amount?: number | null;
  remaining_amount?: number | null;
  security_type?: string | null;
  security_value?: string | null;
  pickup_date?: string | null;
  return_date?: string | null;
}

// ---------------------------------------------------------------------------
// Story 4.2: Owner Order Board types
// ---------------------------------------------------------------------------

export interface PaginationMeta {
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface InternalOrderInput {
  items: { garment_id: string; transaction_type: "buy"; size?: string }[];
  notes?: string;
}

export interface OrderListItem {
  id: string;
  status: OrderStatus;
  payment_status: PaymentStatus;
  subtotal_amount: number;
  discount_amount: number;
  total_amount: number;
  payment_method: PaymentMethod;
  customer_name: string;
  customer_phone: string;
  is_internal?: boolean;
  transaction_types: string[];
  created_at: string;
  next_valid_status: string | null;
  // Epic 10: service type for badge display
  service_type?: ServiceType;
  // Story 10.5: preparation sub-step tracking
  preparation_step?: string | null;
}

// ---------------------------------------------------------------------------
// Story 10.4: Owner Approve & Auto-routing types
// ---------------------------------------------------------------------------

export interface ApproveOrderRequest {
  assigned_to?: string;  // Tailor UUID — required for bespoke
  notes?: string;
}

export interface ApproveOrderResponse {
  order_id: string;
  new_status: string;
  service_type: string;
  routing_destination: string;  // "tailor" | "warehouse"
  task_id?: string | null;
}

// Story 10.5: Preparation sub-step constants
export const RENT_PREP_STEPS = [
  { key: "cleaning", label: "Giặt/Là" },
  { key: "altering", label: "Chỉnh sửa" },
  { key: "ready", label: "Sẵn sàng" },
] as const;

export const BUY_PREP_STEPS = [
  { key: "qc", label: "Kiểm tra CL" },
  { key: "packaging", label: "Đóng gói" },
  { key: "ready", label: "Sẵn sàng" },
] as const;

export interface UpdatePreparationStepRequest {
  preparation_step: string;
  delivery_mode?: "ship" | "pickup";
}

export interface UpdatePreparationStepResponse {
  order_id: string;
  preparation_step: string | null;
  status: string;
  service_type: string;
  is_completed: boolean;
}

export interface OrderListParams {
  status?: OrderStatus[];
  payment_status?: PaymentStatus[];
  transaction_type?: "buy" | "rent" | "bespoke";
  is_internal?: boolean;
  search?: string;
  page?: number;
  page_size?: number;
  sort_by?: "created_at" | "total_amount" | "status";
  sort_order?: "asc" | "desc";
}

export interface OrderListResponse {
  data: OrderListItem[];
  meta: {
    pagination: PaginationMeta;
  };
}

export interface PaymentTransactionItem {
  id: string;
  order_id: string;
  provider: string;
  transaction_id: string;
  amount: number;
  status: string;
  created_at: string;
}

export interface OrderDetailResponse {
  order: OrderResponse;
  transactions: PaymentTransactionItem[];
}

// ---------------------------------------------------------------------------
// Story 4.4c: Customer-facing Order History types
// ---------------------------------------------------------------------------

export type CustomerOrderStatus =
  | "pending"
  | "confirmed"
  | "in_progress"
  | "checked"
  | "shipped"
  | "delivered"
  | "cancelled"
  | "returned"
  | "overdue";

export type OrderType = "buy" | "rental" | "mixed";

export interface CustomerOrderItem {
  garment_id: string;
  garment_name: string;
  image_url: string | null;
  transaction_type: "buy" | "rent";
  size?: string | null;
  quantity: number;
  unit_price: number;
  total_price: number;
  // Rental-specific
  start_date?: string | null;
  end_date?: string | null;
  rental_days?: number | null;
  rental_status?: string | null;
  deposit_amount?: number | null;
}

export interface OrderTimelineEntry {
  status: string;
  timestamp: string;
  description: string;
}

export interface CustomerOrderDeliveryInfo {
  recipient_name: string;
  phone: string;
  address: string;
  notes?: string | null;
}

export interface CustomerOrderSummary {
  id: string;
  order_number: string;
  total_amount: number;
  status: string;
  payment_status: string;
  order_type: OrderType;
  created_at: string;
}

export interface TailorInfoForCustomer {
  full_name: string;
  avatar_url: string | null;
  role: string;
  experience_years: number | null;
  production_step: string;
  garment_name: string;
}

export interface CustomerOrderDetail extends CustomerOrderSummary {
  payment_method: string;
  shipping_note?: string | null;
  items: CustomerOrderItem[];
  delivery_info: CustomerOrderDeliveryInfo;
  timeline: OrderTimelineEntry[];
  tailor_info: TailorInfoForCustomer[] | null;
}

export interface CustomerOrderListMeta {
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface CustomerOrderListResponse {
  data: CustomerOrderSummary[];
  meta: CustomerOrderListMeta;
}

export interface CustomerOrderFilter {
  status?: string;
  order_type?: "buy" | "rental";
  date_from?: string;
  date_to?: string;
  search?: string;
}

File: backend/src/models/order.py
"""Pydantic schemas for Order and OrderItem (Story 3.3)."""

import re
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

_VN_PHONE_RE = re.compile(r"^(0[35789])\d{8}$")


class PaymentMethod(str, Enum):
    """Supported payment methods."""

    cod = "cod"
    vnpay = "vnpay"
    momo = "momo"
    internal = "internal"


class OrderStatus(str, Enum):
    """Order lifecycle statuses — expanded for Epic 10 Unified Order Workflow."""

    # --- Existing statuses (backward compatible) ---
    pending = "pending"
    confirmed = "confirmed"
    in_progress = "in_progress"  # Legacy — kept for backward compatibility
    checked = "checked"          # Legacy — kept for backward compatibility
    shipped = "shipped"
    delivered = "delivered"
    cancelled = "cancelled"
    # --- Epic 10: New statuses by service_type ---
    pending_measurement = "pending_measurement"  # Bespoke only — chờ xác nhận số đo
    preparing = "preparing"                      # Buy/Rent — đang chuẩn bị (sub-steps)
    ready_to_ship = "ready_to_ship"              # Sẵn sàng giao hàng
    ready_for_pickup = "ready_for_pickup"        # Sẵn sàng nhận tại tiệm
    in_production = "in_production"              # Bespoke only — đang trong sản xuất
    renting = "renting"                          # Rent only — khách đang giữ đồ thuê
    returned = "returned"                        # Rent only — khách đã trả đồ
    completed = "completed"                      # Hoàn tất toàn bộ lifecycle


class ServiceType(str, Enum):
    """Service type for Epic 10 Unified Order Workflow."""

    buy = "buy"          # Mua sẵn — thanh toán 100% upfront
    rent = "rent"        # Thuê — Deposit + Security Deposit + Remaining
    bespoke = "bespoke"  # Đặt may — Deposit + Remaining


class SecurityType(str, Enum):
    """Security deposit type for rental orders (Epic 10)."""

    cccd = "cccd"                  # Căn cước công dân (ID card)
    cash_deposit = "cash_deposit"  # Tiền cọc thế chân (cash)


class RentalCondition(str, Enum):
    """Condition of returned rental item (Story 10.7)."""

    good = "Good"        # Tốt — hoàn trả 100% cọc
    damaged = "Damaged"  # Hỏng — hoàn trả 100% (MVP, no damage fee)
    lost = "Lost"        # Thất lạc — không hoàn trả cọc


class PaymentStatus(str, Enum):
    """Payment status for orders and transactions."""

    pending = "pending"
    paid = "paid"
    failed = "failed"
    refunded = "refunded"


# Story 10.5: Preparation sub-step definitions per service type
RENT_PREP_STEPS = ["cleaning", "altering", "ready"]
BUY_PREP_STEPS = ["qc", "packaging", "ready"]


class ShippingAddress(BaseModel):
    """Vietnamese shipping address format."""

    province: str = Field(..., min_length=1, description="Tinh/Thanh pho")
    district: str = Field(..., min_length=1, description="Quan/Huyen")
    ward: str = Field(..., min_length=1, description="Phuong/Xa")
    address_detail: str = Field(..., min_length=5, description="So nha, ten duong")


class RentalCheckoutFields(BaseModel):
    """Rental-specific checkout fields (Story 10.3)."""

    pickup_date: date
    return_date: date
    security_type: SecurityType
    security_value: str = Field(..., min_length=1, max_length=50)

    @field_validator("return_date")
    @classmethod
    def return_after_pickup(cls, v: date, info) -> date:
        if "pickup_date" in info.data and v <= info.data["pickup_date"]:
            raise ValueError("Ngay tra phai sau ngay nhan")
        return v


class OrderItemCreate(BaseModel):
    """Schema for creating an order item."""

    garment_id: UUID
    transaction_type: str = Field(..., pattern=r"^(buy|rent|bespoke)$")
    size: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    rental_days: int | None = Field(None, ge=1)


class OrderCreate(BaseModel):
    """Schema for creating an order."""

    customer_name: str = Field(..., min_length=2, max_length=255)
    customer_phone: str = Field(..., max_length=20)
    shipping_address: ShippingAddress
    shipping_note: str | None = Field(None, max_length=500)
    payment_method: PaymentMethod = PaymentMethod.cod
    items: list[OrderItemCreate] = Field(..., min_length=1)
    voucher_codes: list[str] = Field(default_factory=list)
    # Story 10.3: Service-type checkout fields
    rental_fields: RentalCheckoutFields | None = None
    measurement_confirmed: bool = False

    @field_validator("customer_phone")
    @classmethod
    def validate_phone(cls, v: str) -> str:
        """Validate Vietnamese phone number format."""
        cleaned = v.strip()
        if not _VN_PHONE_RE.match(cleaned):
            msg = "So dien thoai khong hop le (VN format: 0xx xxx xxxx)"
            raise ValueError(msg)
        return cleaned

    @model_validator(mode="after")
    def validate_service_type_requirements(self) -> "OrderCreate":
        """Cross-field validation: rental_fields required for rent, measurement for bespoke."""
        has_rent = any(i.transaction_type == "rent" for i in self.items)
        has_bespoke = any(i.transaction_type == "bespoke" for i in self.items)
        if has_rent and self.rental_fields is None:
            raise ValueError("rental_fields bat buoc cho don thue")
        if has_bespoke and not self.measurement_confirmed:
            raise ValueError("Phai xac nhan so do truoc khi dat may")
        return self


class OrderItemResponse(BaseModel):
    """Schema for order item in response."""

    garment_id: UUID
    garment_name: str
    image_url: str | None = None
    transaction_type: str
    size: str | None = None
    rental_days: int | None = None
    unit_price: Decimal
    total_price: Decimal

    model_config = {"from_attributes": True}


class OrderResponse(BaseModel):
    """Schema for order response."""

    id: UUID
    status: OrderStatus
    payment_status: PaymentStatus = PaymentStatus.pending
    subtotal_amount: Decimal
    discount_amount: Decimal = Decimal("0")
    total_amount: Decimal
    applied_voucher_ids: list[str] = []
    payment_method: PaymentMethod
    payment_url: str | None = None
    customer_name: str
    customer_phone: str
    shipping_address: ShippingAddress | None = None
    shipping_note: str | None = None
    is_internal: bool = False
    items: list[OrderItemResponse] = []
    created_at: datetime
    # Epic 10 fields — defaults ensure backward compatibility with existing orders
    service_type: ServiceType = ServiceType.buy
    deposit_amount: Decimal | None = None
    remaining_amount: Decimal | None = None
    # Story 10.3: Rental/security fields
    security_type: str | None = None
    security_value: str | None = None
    pickup_date: datetime | None = None
    return_date: datetime | None = None
    # Story 10.5: Preparation sub-step tracking
    preparation_step: str | None = None

    model_config = {"from_attributes": True}


class InternalOrderCreate(BaseModel):
    """Schema for creating an internal (Owner production) order."""

    items: list[OrderItemCreate] = Field(..., min_length=1)
    notes: str | None = Field(None, max_length=500)


class PaginationMeta(BaseModel):
    """Pagination metadata for list responses."""

    page: int
    page_size: int
    total: int
    total_pages: int


class OrderFilterParams(BaseModel):
    """Query params for filtering the order list."""

    status: list[OrderStatus] | None = None
    payment_status: list[PaymentStatus] | None = None
    transaction_type: str | None = Field(None, pattern=r"^(buy|rent|bespoke)$")
    is_internal: bool | None = None
    search: str | None = Field(None, max_length=255)
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
    sort_by: str = Field("created_at", pattern=r"^(created_at|total_amount|status)$")
    sort_order: str = Field("desc", pattern=r"^(asc|desc)$")


class OrderListItem(BaseModel):
    """Order summary row for list view."""

    id: UUID
    status: OrderStatus
    payment_status: PaymentStatus
    subtotal_amount: Decimal
    discount_amount: Decimal = Decimal("0")
    total_amount: Decimal
    payment_method: PaymentMethod
    customer_name: str
    customer_phone: str
    is_internal: bool = False
    transaction_types: list[str] = []
    created_at: datetime
    next_valid_status: str | None = None
    # Epic 10: service type for badge display
    service_type: ServiceType = ServiceType.buy
    # Story 10.5: Preparation sub-step tracking
    preparation_step: str | None = None
    # Story 10.6: remaining amount for payment indicator
    remaining_amount: Decimal | None = None

    model_config = {"from_attributes": True}


class OrderListResponse(BaseModel):
    """Response for paginated order list."""

    data: list[OrderListItem]
    meta: PaginationMeta


class OrderStatusUpdate(BaseModel):
    """Request body for status update."""

    status: OrderStatus


class PaymentTransactionResponse(BaseModel):
    """Schema for payment transaction response."""

    id: UUID
    order_id: UUID
    provider: str
    transaction_id: str
    amount: Decimal
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class OrderPaymentRecord(BaseModel):
    """Schema for order_payments table (Epic 10).

    Business-level multi-transaction tracking for deposit/remaining/security payments.
    Separate from PaymentTransactionResponse (webhook audit trail — Story 4.1).
    """

    id: UUID
    tenant_id: UUID
    order_id: UUID
    payment_type: str  # full | deposit | remaining | security_deposit
    amount: Decimal
    method: str        # cod | vnpay | momo | cash | internal
    status: str        # pending | paid | failed | refunded
    gateway_ref: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class MeasurementCheckResponse(BaseModel):
    """Response schema for measurement gate check (Story 10.2).

    Returns whether the customer has valid measurements for bespoke orders.
    """

    has_measurements: bool
    last_updated: datetime | None = None
    measurements_summary: dict[str, float | None] | None = None


class ApproveOrderRequest(BaseModel):
    """Request body for Owner order approval (Story 10.4).

    assigned_to is required for bespoke orders (auto-creates TailorTask).
    """

    assigned_to: UUID | None = None  # Tailor user ID — required for bespoke
    notes: str | None = Field(None, max_length=2000)


class ApproveOrderResponse(BaseModel):
    """Response for order approval with routing destination (Story 10.4)."""

    order_id: UUID
    new_status: str
    service_type: str
    routing_destination: str  # "tailor" | "warehouse"
    task_id: UUID | None = None


# ── Story 10.5: Preparation Sub-step Tracking ────────────────────────────────


class UpdatePreparationStepRequest(BaseModel):
    """Request body for advancing preparation sub-step (Story 10.5).

    preparation_step: target step (forward-only).
    delivery_mode: required when advancing to the last step (triggers status transition).
    """

    preparation_step: str = Field(..., description="Target preparation step (forward-only)")
    delivery_mode: str | None = Field(
        None,
        pattern=r"^(ship|pickup)$",
        description="Required on last step: 'ship' → ready_to_ship, 'pickup' → ready_for_pickup",
    )


class UpdatePreparationStepResponse(BaseModel):
    """Response for preparation step update (Story 10.5)."""

    order_id: UUID
    preparation_step: str | None  # None if completed (moved out of preparing)
    status: str
    service_type: str
    is_completed: bool  # True if order moved out of preparing


# ── Story 10.6: Remaining Payment & Handover ────────────────────────────────


class PayRemainingRequest(BaseModel):
    """Request body for remaining payment initiation (Story 10.6)."""

    payment_method: PaymentMethod = PaymentMethod.vnpay


class PayRemainingResponse(BaseModel):
    """Response for remaining payment initiation (Story 10.6)."""

    order_id: UUID
    payment_url: str | None
    amount: Decimal
    payment_type: str  # "remaining"


# ── Story 10.7: Rental Return & Security Refund ────────────────────────────────


class RefundSecurityRequest(BaseModel):
    """Request body for security deposit refund processing (Story 10.7).

    condition: Rental item condition (Good | Damaged | Lost).
    """

    condition: RentalCondition = Field(..., description="Rental item condition")


class RefundSecurityResponse(BaseModel):
    """Response for security deposit refund processing (Story 10.7)."""

    order_id: UUID
    refund_amount: Decimal  # Calculated based on condition
    security_type: str  # cccd or cash_deposit
    original_amount: str | None  # Original security deposit value (str from DB VARCHAR)
    condition: RentalCondition  # Good | Damaged | Lost
    already_processed: bool = False  # True if this is an idempotent return
